sbtc-core/src/signer/mod.rs

```diff
@@ -0,0 +1,119 @@
+/// An sBTC signer configuration
+pub mod config;
+
+use std::collections::HashMap;
+
+use crate::signer::config::Config;
+use crate::SBTCError;
+use bitcoin::{Network, PrivateKey, PublicKey, Transaction as BitcoinTransaction};
+use p256k1::ecdsa;
+use url::Url;
+use wsts::{bip340::SchnorrProof, common::Signature};
+
+#[derive(Default, Clone, Debug)]
+pub struct SignersData {
+    pub signers: HashMap<u32, ecdsa::PublicKey>,
+    pub vote_shares: HashMap<u32, ecdsa::PublicKey>,
+}
+
+/// Coordinator algorithm trait
+pub trait CoordinatorAlgorithm {
+    fn generate_sbtc_wallet_public_key(&self) -> Result<PublicKey, SBTCError>;
+    fn run_signing_round(
+        &self,
+        _tx: BitcoinTransaction,
+    ) -> Result<(Signature, SchnorrProof), SBTCError>;
+}
+
+/// Signer algorithm trait
+pub trait SignerAlgorithm: CoordinatorAlgorithm {
+    fn sign_message(&self, message: &[u8]) -> Result<Vec<u8>, SBTCError>;
+    fn verify_message(
+        &self,
+        public_key: &ecdsa::PublicKey,
+        message: &[u8],
+    ) -> Result<bool, SBTCError>;
+}
+
+/// sBTC compliant Signer
+pub struct Signer<S: SignerAlgorithm> {
+    /// Signer configuration
+    pub config: Config,
+    /// Signer private key
+    pub private_key: PrivateKey,
+    /// Network to use
+    pub network: Network,
+    /// The stacks node RPC URL
+    pub stacks_node_rpc_url: Url,
+    /// The signing algorithm
+    pub algorithm: S,
+}
+
+impl<S: SignerAlgorithm> Signer<S> {
+    /// Create a new signer
+    pub fn new(
+        config: Config,
+        private_key: PrivateKey,
+        network: Network,
+        stacks_node_rpc_url: Url,
+        algorithm: S,
+    ) -> Self {
+        Self {
+            config,
+            private_key,
+            network,
+            stacks_node_rpc_url,
+            algorithm,
+        }
+    }
+
+    /// Get the current coordinator public key from the stacks node
+    pub fn get_coordinator_public_key(&self) -> Result<PublicKey, SBTCError> {
+        todo!()
+    }
```

Comment:
What is the coordinator public key in this context? I was under the impression that the "coordinator" would be a different signer for each transaction to be signed.

Reply:
You would want to know if the person requesting you to sign is the valid coordinator for the given round or potentially transaction. I was not convinced that it being per transaction makes sense.

Reply:
Sure, I'd imagine that being something that the signer can derive from the information it has. I.e. if it knows the set of signers and which round or tx it regards, it would be able to easily compute who is the coordinator (primary and secondary).